[nrfconnect] Added the Device Appearance attribute to factory data (#106)
ArekBalysNordic/connectedhomeip: docs/guides/nrfconnect_factory_data_configuration.md

|  `discriminator`  |            Discriminator             |    2 B     |    uint16    |  mandatory  |                                                                                                                                                   A 12-bit value matching the field of the same name in the setup code. The discriminator is used during the discovery process.                                                                                                                                                    |
|    `passcode`     |            SPAKE passcode            |    4 B     |    uint32    |  optional   | A pairing passcode is a 27-bit unsigned integer which serves as a proof of possession during the commissioning. Its value must be restricted to the values from `0x0000001` to `0x5F5E0FE` (`00000001` to `99999998` in decimal), excluding the following invalid passcode values: `00000000`, `11111111`, `22222222`, `33333333`, `44444444`, `55555555`, `66666666`, `77777777`, `88888888`, `99999999`, `12345678`, `87654321`. |

> |   `product_appearance`    | Product visible appearance | 2 B | JSON string | optional | The appearance field is a structure that describes the visible appearance of the product. This field is provided in the JSON format and consist of two attributes: `finish` (1B), `primary_color` (1B). See the [The appearance field description](#the-appearance-field-description) to learn how to set all attributes. | 

Reviewer: Does this table describe the intermediate JSON format or the factory data on the device? If the latter (which I think is the case based on the previous rows), it should be "CBOR map" rather than "JSON string". Same applies to the row below.

> |   `appearance`    | Device appearance | 5 B | JSON string | optional | A appearance filed is a structure which describes the visible appearance of the product. This field is provided in the JSON format and consist of four attributes: `finish` (1B), `primary_color_hue` (2B), `primary_color_saturation` (1B), and `primary_color_value` (1B). See the [The appearance field description](#the-appearance-field-description) to learn how to set all attributes. | 

Reviewer 2: ```suggestion
|   `appearance`    | Device appearance | 5 B | JSON string | optional | An appearance field is a structure which describes the visible appearance of the product. This field is provided in the JSON format and consist of four attributes: `finish` (1B), `primary_color_hue` (2B), `primary_color_saturation` (1B), and `primary_color_value` (1B). See the [The appearance field description](#the-appearance-field-description) to learn how to set all attributes. | 
```

> |   `appearance`    | Device appearance | 5 B | JSON string | optional | A appearance filed is a structure which describes the visible appearance of the product. This field is provided in the JSON format and consist of four attributes: `finish` (1B), `primary_color_hue` (2B), `primary_color_saturation` (1B), and `primary_color_value` (1B). See the [The appearance field description](#the-appearance-field-description) to learn how to set all attributes. | 

Reviewer 2: How about calling it `product_appearance` to precisely match spec attribute name?

|  `discriminator`  |            Discriminator             |    2 B     |    uint16    |  mandatory  |                                                                                                                                                   A 12-bit value matching the field of the same name in the setup code. The discriminator is used during the discovery process.                                                                                                                                                    |
|    `passcode`     |            SPAKE passcode            |    4 B     |    uint32    |  optional   | A pairing passcode is a 27-bit unsigned integer which serves as a proof of possession during the commissioning. Its value must be restricted to the values from `0x0000001` to `0x5F5E0FE` (`00000001` to `99999998` in decimal), excluding the following invalid passcode values: `00000000`, `11111111`, `22222222`, `33333333`, `44444444`, `55555555`, `66666666`, `77777777`, `88888888`, `99999999`, `12345678`, `87654321`. |

> |`product_appearance`|     Product visible appearance      |    2 B     | CBOR map | optional | The appearance field is a structure that describes the visible appearance of the product. This field is provided in the CBOR map and consist of two attributes: `finish` (1B), `primary_color` (1B). See the [The appearance field description](#the-appearance-field-description) to learn how to set all attributes. | 

Reviewer 3: ```suggestion
|`product_appearance`|     Product visible appearance      |    2 B     | CBOR map | optional | The appearance field is a structure that describes the visible appearance of the product. This field is provided in a CBOR map and consists of two attributes: `finish` (1 B), `primary_color` (1 B). See the [Appearance field description](#appearance-field-description) to learn how to set all attributes. | 
```
To remove double "the". Also, we we don't usually need "the" in headers.

> |      `user`       |              User data               |  variable, max 1024 B  | CBOR map  | optional  |                             The user data is provided in the JSON format. This parameter is optional and depends on device manufacturer's purpose. It is provided as a CBOR map type from persistent storage and should be parsed in the user application. This data is not used by the Matter stack. To learn how to work with user data, see [How to set user data](#how-to-set-user-data) section.                              |

Reviewer 3: ```suggestion
|      `user`       |              User data               |  variable, max 1024 B  | CBOR map  | optional  | The user data is provided in the JSON format. This parameter is optional and depends on the device manufacturer's purpose. It is provided as a CBOR map type from persistent storage and should be parsed in the user application. This data is not used by the Matter stack. To learn how to work with user data, see the [How to set user data](#how-to-set-user-data) section.                              |
```

-   All certificates stored in factory data are provided in the
    [X.509](https://www.itu.int/rec/T-REC-X.509-201910-I/en) format.

#### Appearance field description

The `appearance` field in the factory data set describes the device's visible appearance.

- `finish` - A string name that indicates the visible exterior finish of the product. 
It refers to the `ProductFinishEnum` enum, and currently, you can choose one of the following names:

|   Name     | Enum value |
| :--------: | :--------: |
| `matte`    |     0      |
| `satin`    |     1      |
| `polished` |     2      |
| `rugged`   |     3      |
| `fabric`   |     4      |
| `other`    |     255    |

- `primary_color` - A string name that represents the RGB color space of the device's case color, which is the most representative. 
It refers to the `ColorEnum` enum, and currently, you can choose one of the following names:

(Enum value) color name (`RGB value`)

> | (0) $$\color{black} \fcolorbox{white}{white}{black}$$ (`#000000`) | (1) $$\color{#000080}{navy}$$ (`#000080`) | (2) $$\color{#008000}{green}$$ (`#008000`)| (3) $$\color{#008080}{teal}$$ (`#008080`) | (4) $$\color{#800080}{maroon}$$ (`#800080`) |

Reviewer 3: Something wrong in the preview:
![image](https://github.com/ArekBalysNordic/connectedhomeip/assets/41291385/b206b84c-275e-4305-bd46-78e155eb5b5b)

Author: Repaired.

-   All certificates stored in factory data are provided in the
    [X.509](https://www.itu.int/rec/T-REC-X.509-201910-I/en) format.

#### The appearance field description

The `appearance` field in the factory data set describes device's visible appearance and it consist of the following attributes:

- `finish` - This field is a uint8 number and indicates the visible finish of the product. 
It refers to the `ProductFinishEnum` enum, and currently, it can be set in the following way:
    - 0 - the product has a matte finish.
    - 1 - the product has a satin finish.
    - 2 - the product has a shiny or polished finish.

> - `primary_color_hue` - This field is the 16-bit integer and it represents the HSV color space hue of the device's color, which is the most representative. 

Reviewer 2: How did you get information about this saturation, hue and etc? I saw the PR to spec, but there is only ColorEnum that describes primary color in RGB. I didn't found any information about the sub-fields.

The value is rounded to the nearest degree of the HSV color circle and the 360 value is represented as 0.

> - `primary_color_saturation` - This field is a percentage rounded to the nearest value of the HSV color space saturation. 
> The value 0 means the least saturated color, and the value 100 means the most saturated color
> The chosen color should be the most representative and most of the product. 

Reviewer 3: ```suggestion
- `primary_color_saturation` - A percentage value rounded to the nearest value of the HSV color space saturation. 
The value `0` means the least saturated color, and the value `100` means the most saturated color.
The chosen color should be the most representative for the product and the most dominant.
```

> - `primary_color_value` - This field is a percentage rounded to the nearest value of the HSV color space value. 
> The value 0 means the darkest color, and the value 100 means the lightest color.
> The chosen color should be the most representative of the product.

Reviewer 3: ```suggestion
- `primary_color_value` - A percentage value rounded to the nearest value of the HSV color space value. 
The value `0` means the darkest color, and the value `100` means the lightest color.
The chosen color should be the most representative of the product.
```

-   All certificates stored in factory data are provided in the
    [X.509](https://www.itu.int/rec/T-REC-X.509-201910-I/en) format.

> #### The appearance field description

Reviewer 3: ```suggestion
#### Appearance field description
```

Reviewer 3: This edit is valid only for this header.

The `appearance` field in the factory data set describes the device's visible appearance.

- `finish` - A string name that indicates the visible exterior finish of the product. 
It refers to the `ProductFinishEnum` enum, and currently, you can choose one of the following names:

|   Name     | Enum value |
| :--------: | :--------: |
| `matte`    |     0      |
| `satin`    |     1      |
| `polished` |     2      |
| `rugged`   |     3      |
| `fabric`   |     4      |
| `other`    |     255    |

- `primary_color` - A string name that represents the RGB color space of the device's case color, which is the most representative. 
It refers to the `ColorEnum` enum, and currently, you can choose one of the following names:

(Enum value) color name (`RGB value`)

| (0) $$\color{black} \fcolorbox{white}{white}{black}$$ (`#000000`) | (1) $$\color{#000080}{navy}$$ (`#000080`) | (2) $$\color{#008000}{green}$$ (`#008000`)| (3) $$\color{#008080}{teal}$$ (`#008080`) | (4) $$\color{#800080}{maroon}$$ (`#800080`) |
| --------------------- | ----------------- | --------------- | ----------------- | --------------- |

> | (5) $$\color{#800080}{purple}$$ (`#800080`) | (6) $$\color{#808000}{olive}$$ (`#800080`) | (7) $$\color{#808080}{gray}$$ (`#800080`) | (8) $$\color{#0000FF}{blue}$$ (`#0000FF`) | (9) $$\color{#00FF00}{lime}$$ (`#00FF00`) |

Reviewer: Nice!

--discriminator 0xF00 \
--generate_rd_uid \
--passcode 20202021 \

> --appearance '{"finish": "satin", "primary_color": "green"}' \

Reviewer: Needs fixing.
